include/hip/spirv_hip_fp16.h

```diff
@@ -18,1011 +18,1713 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-    data = x.data;
-    return *this;
-  }
-  __host__ __device__ volatile __half& operator=(const __half_raw& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(const volatile __half_raw& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  __half& operator=(__half_raw&& x) {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(__half_raw&& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(volatile __half_raw&& x) volatile {
-    data = x.data;
-    return *this;
-  }
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_floating_point<T>{}>* = nullptr>
-  __host__ __device__ __half& operator=(T x) {
-    data = static_cast<_hip_f16>(x);
-    return *this;
-  }
-#endif
+#define __ocml_exp10_f16(__X) exp10_h(__X)
+#define __ocml_exp10_2f16(__X) exp10_2h(__X)
 
-// MANIPULATORS - DEVICE ONLY
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_integral<T>{}>* = nullptr>
-  __device__ __half& operator=(T x) {
-    data = static_cast<_hip_f16>(x);
-    return *this;
-  }
-#endif
+#define __ocml_log2_f16(__X) log2_h(__X)
+#define __ocml_log2_2f16(__X) log2_2h(__X)
 
-#if !defined(__HIP_NO_HALF_OPERATORS__)
-  __device__ __half& operator+=(const __half& x) {
-    data += x.data;
-    return *this;
-  }
-  __device__ __half& operator-=(const __half& x) {
-    data -= x.data;
-    return *this;
-  }
-  __device__ __half& operator*=(const __half& x) {
-    data *= x.data;
-    return *this;
-  }
-  __device__ __half& operator/=(const __half& x) {
-    data /= x.data;
-    return *this;
-  }
-  __device__ __half& operator++() {
-    ++data;
-    return *this;
-  }
-  __device__ __half operator++(int) {
-    __half tmp{*this};
-    ++*this;
-    return tmp;
-  }
-  __device__ __half& operator--() {
-    --data;
-    return *this;
-  }
-  __device__ __half operator--(int) {
-    __half tmp{*this};
-    --*this;
-    return tmp;
-  }
-#endif
+#define __ocml_log_f16(__X) log_h(__X)
+#define __ocml_log_2f16(__X) log_2h(__X)
 
-// ACCESSORS
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_floating_point<T>{} &&
-                                    !std::is_same<T, double>{}>* = nullptr>
-  operator T() const {
-    return data;
-  }
-#endif
-  __host__ __device__ operator __half_raw() const { return __half_raw{data}; }
-  __host__ __device__ operator volatile __half_raw() const volatile {
-    return __half_raw{data};
-  }
-
-// ACCESSORS - DEVICE ONLY
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_integral<T>{}>* = nullptr>
-  __device__ operator T() const {
-    return data;
-  }
+#define __ocml_log10_f16(__X) log10_h(__X)
+#define __ocml_log10_2f16(__X) log10_2h(__X)
+
+#define __llvm_amdgcn_rcp_f16(__X) ((__half)1.0f / (__half)__X)
+#define __llvm_amdgcn_rcp_2f16(__X) ((__half2)1.0f / (__half2)__X)
```

Comment:
Aren't `amdgcn` built-in LLVM intrinsics for AMD? Why do we have any mention of them on the SPIR-V side? @pjaaskel

Reply:
Just for source-level compatibility with the roc libraries which call them directly (in this case rocPRIM). IMHO we can just as well add them to avoid needing to redefine them when porting roc/hip libs which refer to them.